rust/agama-network/src/nm/dbus.rs

```diff
@@ -0,0 +1,443 @@
+//! This module implements some functions to convert from/to D-Bus types
+//!
+//! Working with hash maps coming from D-Bus is rather tedious and it is even worse when working
+//! with nested hash maps (see [NestedHash] and [OwnedNestedHash]).
+use super::model::*;
+use crate::model::*;
+use std::collections::HashMap;
+use std::net::Ipv4Addr;
+use uuid::Uuid;
+use zbus::zvariant::{self, Value};
+
+const ETHERNET_KEY: &str = "802-3-ethernet";
+const WIRELESS_KEY: &str = "802-11-wireless";
+const WIRELESS_SECURITY_KEY: &str = "802-11-wireless-security";
+const LOOPBACK_KEY: &str = "loopback";
+
+/// Nested hash to send to D-Bus.
+type NestedHash<'a> = HashMap<&'a str, HashMap<&'a str, zvariant::Value<'a>>>;
+/// Nested hash as it comes from D-Bus.
+type OwnedNestedHash = HashMap<String, HashMap<String, zvariant::OwnedValue>>;
+
+/// Converts a connection struct into a HashMap that can be sent over D-Bus.
+///
+/// * `conn`: Connection to cnvert.
+pub fn connection_to_dbus(conn: &Connection) -> NestedHash {
+    let mut result = NestedHash::new();
+    let mut connection_dbus =
+        HashMap::from([("id", conn.id().into()), ("type", ETHERNET_KEY.into())]);
+    result.insert("ipv4", ipv4_to_dbus(conn.ipv4()));
+
+    if let Connection::Wireless(wireless) = conn {
+        connection_dbus.insert("type", "802-11-wireless".into());
+        let wireless_dbus = wireless_config_to_dbus(wireless);
+        for (k, v) in wireless_dbus {
+            result.insert(k, v);
+        }
+    }
+    result.insert("connection", connection_dbus);
+    result
+}
+
+/// Converts an OwnedNestedHash from D-Bus into a Connection.
+///
+/// This functions tries to turn a OwnedHashMap coming from D-Bus into a Connection.
+pub fn connection_from_dbus(conn: OwnedNestedHash) -> Option<Connection> {
+    let base = base_connection_from_dbus(&conn)?;
+
+    if let Some(wireless_config) = wireless_config_from_dbus(&conn) {
+        return Some(Connection::Wireless(WirelessConnection {
+            base,
+            wireless: wireless_config,
+        }));
+    }
+
+    if conn.get(LOOPBACK_KEY).is_some() {
+        return Some(Connection::Loopback(LoopbackConnection { base }));
+    };
+
+    if conn.get(ETHERNET_KEY).is_some() {
+        return Some(Connection::Ethernet(EthernetConnection { base }));
+    };
+
+    None
+}
+
+/// Merges a NestedHash and an OwnedNestedHash connections.
+///
+/// * `original`: original hash coming from D-Bus.
+/// * `updated`: updatde hash to write to D-Bus.
+pub fn merge_dbus_connections<'a>(
+    original: &'a OwnedNestedHash,
+    updated: &'a NestedHash,
+) -> NestedHash<'a> {
+    let mut merged = HashMap::new();
+    for (key, orig_section) in original {
+        let mut inner: HashMap<&str, zbus::zvariant::Value> = HashMap::new();
```

Comment:
as it is O^2 complexity, I would at least try to reduce reallocation with usng for both HashMap `::with_capacity` set to original, respective original_section which should eliminaty expensive reallacation.